edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/RegistryOnBehalfOfApiClient.cs

```diff
@@ -0,0 +1,167 @@
+        readonly string listModulesOnBehalfOfUriTemplate = $"devices/{{0}}/modules/{Constants.EdgeHubModuleId}/getModulesOnTargetDevice?api-version={{1}}";
+        readonly string deleteModuleOnBehalfOfUriTemplate = $"devices/{{0}}/modules/{Constants.EdgeHubModuleId}/deleteModuleOnBehalfOf?api-version={{1}}";
+
+        readonly Option<IWebProxy> proxy;
+        readonly ITokenProvider tokenProvider;
+        readonly Uri baseUri;
+
+        public RegistryOnBehalfOfApiClient(string upstreamHostname, Option<IWebProxy> proxy, ITokenProvider tokenProvider)
+        {
+            Preconditions.CheckNonWhiteSpace(upstreamHostname, nameof(upstreamHostname));
+            this.proxy = proxy;
+            this.tokenProvider = Preconditions.CheckNotNull(tokenProvider, nameof(tokenProvider));
+
+            this.baseUri = new UriBuilder(Uri.UriSchemeHttps, upstreamHostname).Uri;
+        }
+
+        public async Task<RegistryApiHttpResult> PutModuleAsync(string actorEdgeDeviceId, CreateOrUpdateModuleOnBehalfOfData requestData, string eTag)
+        {
+            var requesteUri = new Uri(this.baseUri, string.Format(CultureInfo.InvariantCulture, this.putModuleOnBehalfOfUriTemplate, actorEdgeDeviceId, ApiVersion));
+            var response = await this.SendRequestAsync(requesteUri, HttpMethod.Put, requestData, eTag);
+            return response;
+        }
+
+        public async Task<RegistryApiHttpResult> GetModuleAsync(string actorEdgeDeviceId, GetModuleOnBehalfOfData requestData)
+        {
+            var requesteUri = new Uri(this.baseUri, string.Format(CultureInfo.InvariantCulture, this.getModuleOnBehalfOfUriTemplate, actorEdgeDeviceId, ApiVersion));
+            var response = await this.SendRequestAsync(requesteUri, HttpMethod.Post, requestData);
+            return response;
+        }
+
+        public async Task<RegistryApiHttpResult> ListModulesAsync(string actorEdgeDeviceId, ListModulesOnBehalfOfData requestData)
+        {
+            var requesteUri = new Uri(this.baseUri, string.Format(CultureInfo.InvariantCulture, this.listModulesOnBehalfOfUriTemplate, actorEdgeDeviceId, ApiVersion));
+            var response = await this.SendRequestAsync(requesteUri, HttpMethod.Post, requestData);
+            return response;
+        }
+
+        public async Task<RegistryApiHttpResult> DeleteModuleAsync(string actorEdgeDeviceId, DeleteModuleOnBehalfOfData requestData)
+        {
+            var requesteUri = new Uri(this.baseUri, string.Format(CultureInfo.InvariantCulture, this.deleteModuleOnBehalfOfUriTemplate, actorEdgeDeviceId, ApiVersion));
+            var response = await this.SendRequestAsync(requesteUri, HttpMethod.Post, requestData);
+            return response;
+        }
+
+        async Task<RegistryApiHttpResult> SendRequestAsync<T>(Uri uri, HttpMethod method, T payload, string ifMatchHeader = null)
```

Comment:
Nit: Can we use consistent naming for this and call it etag or ifMatchHeader everywehre?